File: app/src/main/java/com/matchthree/game/engine/Match.kt
package com.matchthree.game.engine

import com.matchthree.game.model.Position

/**
 * A maximal horizontal or vertical run of 3+ same-type gems.
 * Shape precedence (5-run > T/L > 4-run > 3-run) is Milestone 4.
 */
data class Match(val positions: List<Position>) {
    init {
        require(positions.size >= 3) { "match must cover at least 3 cells" }
    }

    val size: Int get() = positions.size
}

File: Engine/Rules/Match.cs
using MatchThree.Engine.Model;

namespace MatchThree.Engine.Rules;

public sealed record Match
{
    public List<Position> Positions { get; }

    public int Size => Positions.Count;

    public Match(List<Position> positions)
    {
        if (positions.Count < 3)
            throw new ArgumentException("match must cover at least 3 cells", nameof(positions));
        Positions = positions;
    }
}